[FIT] Support generation of slurm configuration files when using multi-instance type Compute Resources

EddyMM opened the pull request:
### Description of changes
- Slurm configuration files are generated for each queue in the cluster
- The configuration files are made on the assumption that each node maps to one instance type
- This commit adds support for generating Slurm Configuration files for queues that have multiple instance types per Compute Resource
- Assumptions made:
    - All instance types associated with the compute resource have the same number of CPUs and GPUs
    - RealMemory is not supported for Queues with multiple instance types per Compute Resource

### Tests
- Unit tests:
    - Queues with multiple instance types in any Compute Resource:
        - Will not set RealMemory property in the nodes
        - Will not have the instance type as a feature in the nodes

### References
* [Add configuration parameters needed to specify multiple instance types for a compute resource](https://github.com/aws/aws-parallelcluster/pull/4250)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/common/slurm_utils.conf:
```diff
@@ -0,0 +1,7 @@
+
+{%- macro node_name(queue_name, compute_resource_name, size, is_static) -%}
+    {{ queue_name }}-{{ "st" if is_static else "dy" }}-{{ compute_resource_name }}-[1-{{ size }}]
+{%- endmacro -%}
```
add newline at the end of the file

lukeseawalker replied:
restore new line

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/common/slurm_utils.conf:
```diff
@@ -0,0 +1,7 @@
+{%- macro node_name_section(queue_name, compute_resource_name, size, is_static) -%}
```
what about naming the file `slurm_parallelclulster_utils.conf`, so that it's aligned with other files?

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
-    {% if static_size > 0 %}
-NodeName={{ queue_name }}-st-{{ compute_resource.Name }}-[1-{{ static_size }}] CPUs={{ compute_resource | vcpus }} RealMemory={{ compute_resource | realmemory }} State=CLOUD Feature=static,{{ instance_type }},{{ compute_resource.Name }}{% if compute_resource.Efa.Enabled %},efa{% endif %}{% if instance_type | gpus > 0 %},gpu{% if not no_gpu %} Gres=gpu:{{ instance_type | gpu_type }}:{{ instance_ty ...
+{#- Macros: #}
+{%- macro cpus(compute_resource) -%}
```
why not moving all these macros into the utils.conf file?

EddyMM replied:
I moved only macros that are common between the queue/partition & queue/gres configuration files.

I'll go ahead and move these macros as well to the utils file to improve the readability of the actual configuration files.

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
+{%- endmacro %}
+
+{%- macro real_memory(compute_resource) %}
+    {%- if not compute_resource.InstanceTypeList -%}
```
#TODO single instance type in InstanceTypeList has RealMemory set

EddyMM replied:
Done ([real_memory_property](https://github.com/aws/aws-parallelcluster-cookbook/pull/1499/files#diff-6668f338ff431ea63de7cd7dd41034a6d93cf95510f2667048c366eb12fbbbbfR17)).

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
+{% macro node_definition(queue_name, compute_resource, static_size, dynamic_size, state="CLOUD") %}
+    {% if static_size > 0 %}
+        {{- slurm_utils.node_name_section(queue_name, compute_resource.Name, static_size, True) -}}
+        {{- cpus(compute_resource) -}}
```
I would be consistent in the naming of these macro. For example

```
node_name_section
cpus_section
...
```

or alternatively use `_setting` or `_property` as suffix. What's the name convention that Slurm uses for these?

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
+    {% if dynamic_size > 0 %}
+        {{- slurm_utils.node_name_section(queue_name, compute_resource.Name, dynamic_size, False) -}}
+        {{- cpus(compute_resource) }}{{ real_memory(compute_resource) -}}
+        {{- node_state(state) -}}
```
just `state`+ suffix

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
+
+{%- macro feature(compute_resource, is_static) -%}
+    {{ ' ' }}Feature={{ "static" if is_static else "dynamic" }}
+    {%- if not compute_resource.InstanceTypeList -%}
```
#TODO for single instance type in InstanceTypeList we can set the instance type

EddyMM replied:
Done([feature_property](https://github.com/aws/aws-parallelcluster-cookbook/pull/1499/files#diff-6668f338ff431ea63de7cd7dd41034a6d93cf95510f2667048c366eb12fbbbbfR25)).

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
-        {% set ns.has_node = true %}
+    {% if dynamic_size > 0 %}
+        {{- slurm_utils.node_name_section(queue_name, compute_resource.Name, dynamic_size, False) -}}
+        {{- cpus(compute_resource) }}{{ real_memory(compute_resource) -}}
```
missing new line for `{{ real_memory....`

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_queue_partition.conf:
```diff
@@ -1,38 +1,89 @@
 PartitionName={{ queue_name }} Nodes={{ queue_name }}_nodes MaxTime=INFINITE State=UP{% if is_default_queue %} Default=YES{% endif %}
 
-{% endif %}
+{% endif %}
```
restore the new line

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:
```diff
@@ -199,6 +200,17 @@ def _get_jinja_env(template_directory, realmemory_to_ec2memory_ratio):
     return env
 
 
+def _instance_type_from_compute_resource(compute_resource):
```
what about just `_instance_type`?

lukeseawalker commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:
```diff
@@ -228,7 +240,7 @@ def _gpu_type(instance_type):
 def _vcpus(compute_resource) -> int:
     """Get the number of vcpus for the instance according to disable_hyperthreading and instance features."""
-    instance_type = compute_resource["InstanceType"]
+    instance_type = _instance_type_from_compute_resource(compute_resource)
```
#TODO use this also in `def _realmemory` so to account single instance type in InstanceTypeList

EddyMM committed:
[FIT] Support generation of slurm configuration files when using multi-instance type Compute Resources

- Slurm configuration files are generated for each queue in the cluster
- The configuration files are made on the assumption that each node maps to one instance type
- This commit adds support for generating Slurm Configuration files for queues that have multiple instance types per Compute Resource
- Assumptions made:
    - All instance types associated with the compute resource have the same number of CPUs and GPUs

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>
